lib/cache.cpp

```diff
@@ -0,0 +1,296 @@
+#include "cache.h"
+#include "errorlogger.h"
+#include <cryptopp/sha3.h>
+#include <cryptopp/base64.h>
+#include <cryptopp/files.h>
+#include <fstream>
+using namespace std;
+using namespace tinyxml2;
+using namespace CryptoPP;
+
+Cache::Cache()
+{
+}
+
+Cache::Cache(const Cache & other)
+{
+	*this = other;
+}
+
+Cache & Cache::operator=(const Cache & other)
+{
+	Clear();
+	if (!_cacheFile.empty())
+	{
+		Load(other._cacheFile);
+	}
+	return *this;
+}
+
+int Cache::Load(const string& file)
+{
+	XMLError er = XML_SUCCESS;
+
+	if (!_cacheFile.empty())
+	{
+		return -1;
+	}
+
+	_cacheFile = file;
+	er = _cache.LoadFile(file.c_str());
+
+	return((er == XML_SUCCESS) || (er == XML_NO_ERROR)) ? 0 : (er | XML_SUCCESS);
+}
+
+int Cache::LoadFromCache(const char* filePath, const char* configuration, size_t* pCachedSize, std::string* pHash, const tinyxml2::XMLElement** ppElem) const
+{
+	const XMLElement *elem = NULL;
+	const char* sizeStr = NULL;
+	const char* hashStr = NULL;
+	size_t tmpSize = 0;
+	string path;
+	(*pCachedSize) = 0;
+	(*ppElem) = NULL;
+	pHash->clear();
+
+	path = Normalize(filePath);
+
+	if ((Find(path.c_str(), configuration, &elem) != 0) || (elem == NULL))
+	{
+		return -1;
+	}
+
+	// Parse size
+	if ((sizeStr = elem->Attribute("Size")) == NULL)
+	{
+		return -1;
+	}
+
+	if ((tmpSize = ::stoull(sizeStr, NULL)) == 0)
+	{
+		return -1;
+	}
+
+	// Hash
+	if ((hashStr = elem->Attribute("Hash")) == NULL)
+	{
+		return -1;
+	}
+
+	(*pCachedSize) = tmpSize;
+	(*pHash) = hashStr;
+	(*ppElem) = elem;
+	return 0;
+}
+
+int Cache::Find(const char* filePath, const char* configuration, XMLElement** ppElem)
+{
+	return Find(filePath, configuration, (const XMLElement**)ppElem);
+}
+
+int Cache::Find(const char* filePath, const char* configuration, const XMLElement** ppElem) const
+{
+	(*ppElem) = NULL;
+
+	string path = Normalize(filePath);
+	if (path.empty() || _cacheFile.empty())
+	{
+		return -1;
+	}
+
+	const XMLElement* root = _cache.RootElement();
+	if (root == NULL)
+	{
+		return -1;
+	}
+
+	for (const XMLElement* currElem = root->FirstChildElement("File"); currElem != NULL; currElem = currElem->NextSiblingElement("File"))
+	{
+		if (currElem->Attribute("Path", path.c_str()) && currElem->Attribute("Configuration", configuration))
+		{
+			(*ppElem) = currElem;
+			break;
+		}
+	}
+
+	return ((*ppElem) == NULL);
+}
+
+string Cache::CalcHash(const char* code) const
+{
+	SHA3_512 sha;
+	string out;
+	Base64Encoder* enc = new Base64Encoder(new StringSink(out), false);
+
+	StringSource(code, true, new HashFilter(sha, enc));
+
+	return out;
+}
+
+string Cache::Normalize(const char* filePath) const
```

Comment:
I think you should use our Path class to deal with filepaths. There is for instance a Path::fromNativeSeparators() function that might be useful here.